### ISSUE TITLE
Future of automated data collection

### ISSUE THREAD

**Issue by jeanpaulrsoucy** [ccodwg/CovidTimelineCanada#6]
# Current update process for `Covid19Canada`

Automated nightly updates of the [legacy dataset](https://github.com/ccodwg/Covid19Canada) (`Covid19Canada`) are presently handled using the following tools:

- The master version of the dataset for each value is stored in `Google Sheets`
- Manual edits for data that cannot be automated or for historical edits to the dataset are done in `Google Sheets`
- [`Covid19CanadaBot`](https://github.com/ccodwg/Covid19CanadaBot), a set of GitHub Actions calling [`Covid19CanadaETL`](https://github.com/ccodwg/Covid19CanadaETL) and [other miscellaneous scripts](https://github.com/ccodwg/Covid19Canada/tree/master/scripts) from `Covid19Canada`
- `Covid19CanadaETL` collates the data after calling [`Covid19CanadaData`](https://github.com/ccodwg/Covid19CanadaData) to download the datasets and [`Covid19CanadaDataProcess`](https://github.com/ccodwg/Covid19CanadaDataProcess) to process the datasets and periodically pushes them to `Google Sheets`
- `Covid19CanadaDataProcess` processes datasets given a UUID and specified value (e.g., `cases`, `mortality`, `vaccine_total_doses`, etc.) and format (i.e., time series or current cumulative value)
- `Covid19CanadaData` downloads datasets identified by their UUID in [`datasets.json`](https://github.com/ccodwg/Covid19CanadaArchive/blob/master/datasets.json) (which stores information about the URL of each dataset)
- [Miscellaneous scripts](https://github.com/ccodwg/Covid19Canada/tree/master/scripts) from the `Covid19Canada` repository handle data the steps of the nightly data update and various validation steps
- The final `Covid19Canada` dataset is currently available through [GitHub](https://github.com/ccodwg/Covid19Canada), our [Flask API](https://opencovid.ca/api/) and [Google Drive](https://drive.google.com/drive/folders/1He6mPAbolgh7jtsq1zu6LpLQKz34n_nP)

The current update process is summarized in the following flowchart:

<img src="https://user-images.githubusercontent.com/31170087/152268749-df6e684b-4570-4374-a85a-19e85b9e6503.png" alt="Flowchart illustrating the update process for Covid19CanadaArchive" width="60%" height="60%">

`Covid19CanadaBot` sends emails when it detects a failure in the update process, such as blank values, unexpected 0s or large changes to cumulative values, all of which typically indicate that the current data source has been modified or retired, requiring an update to `Covid19CanadaDataETL` or `Covid19CanadaDataProcess`.

# Future update process for `CovidTimelineCanada`

`CovidTimelineCanada` may be considered the successor the legacy `Covid19Canada` dataset. It also offers us an opportunity to rethink the automation process. Enter [`kamu`](https://www.kamu.dev/) (and its hosted platform, `Kamu Hub`), which is a tool for building reproducible, verifiable data pipelines.

## A new data pipeline with `kamu`?

My initial thoughts on the future data pipeline and where `kamu` may fit in:

- Data acquisition
  - Where should `kamu pull` point to for the various input data sources? Currently, the data update relies on the following formats: csv, json, xlsx, html (both webpages requiring and not requiring Javascript rendering). While `kamu` can handle csv and json natively, I'm not sure about the latter two, especially if Javascript rendering is required (i.e., Selenium). This is presently handled through `Chrome` in `Docker` [as implemented in `Covid19CanadaData`](https://github.com/ccodwg/Covid19CanadaData/blob/master/R/webdriver.R). Thus, I see the following two options:
    - Pull directly from input data sources for csv and json data and pull from pre-processed csv output data for datasets in xlsx and html format (these could be regularly uploaded to S3 by `Covid19CanadaBot` or something)
    - Pull all data from pre-processed csv output data (i.e., as generated by `Covid19CanadaDataProcess`)
- Data processing
  - As above, should we re-write data processing for csv and json files using `kamu` or simply continue to maintain the already-written data processing in `Covid19CanadaProcess`?
- Data collation
  - `kamu` has a FilesGlob source that can track a folder and ingest every new file that appears in it (e.g., csv source files)
  - `kamu` would be used to aggregate data and calculate derived values (e.g., deltas, 7-day averages, etc.)
- Data verification
  - `Covid19CanadaBot` currently scans the Google Sheets periodically for anomalous values as well as emailing errors arising during the update process. Could this be replaced by a series of SQL queries in `kamu` that continuously checks the data? Could these trigger notifications for manual verification?
- Serving data
  - "Definitive" dataset would be the `kamu` dataset hosted on S3
  - CSV version of dataset mirrored on GitHub and Google Drive
  - How would `kamu` integrate with our existing Flask API? Currently, the API uses `pandas` in Python to return processed data, could this be modified to queries of the `kamu` dataset using SQL?
- Data dashboard/data visualization
  - I like the idea of using stream processing and having two "versions" of the dataset, e.g., to deal with "spikes", weekend non-reporting and related issues. Thus, there would be a "live" version of the dataset with all of the newest available data and a "delayed" version of the dataset that would only include (relatively) complete data for all provinces/territories (e.g., delayed during weekends/holidays).
  - Obviously, the `kamu` setup would flow very nicely into an updated data dashboard. But see also the question about regarding our API.
- Manual data entry
  - What would be the best way to deal with data that must be manually entered? Likely, Google Sheets still has a role for these datasets.
  - Would also have to think about how historical data editing would work for datasets not provided in the form of time series (e.g., datasets that we constructed by capturing the cumulative reported value each day). Again, it is possible Google Sheets has a role here.

**Comment by jeanpaulrsoucy** [comment]
Tagging @sergiimk and @onyalcin.

**Comment by sergiimk** [comment]
## Data ingestion
What do you think about this data flow:
![image](https://user-images.githubusercontent.com/204914/152665782-65dad562-27b3-437e-b272-018ae1fd541e.png)

Source code of the diagram (uses PlantUML): [pipeline.puml](https://github.com/ccodwg/CovidTimelineCanada/files/8009307/pipeline.txt)

Key idea being that data from all sources is first preserved in its close to the original form in S3, and then ingested into `kamu`. This will allow you at any point in time to delete everything else and reconstruct derivative data from scratch - a nice to have peace of mind against any errors or in case we need to make backwards-incompatible changes to `kamu`.

## Data verification
This is a very interesting topic for me - for a very long time I wanted to add ability to write unit tests for data (in a similar style to [dbt tests](https://docs.getdbt.com/docs/building-a-dbt-project/tests)), so this is something we could try to do.

I think key purpose of such tests would be to prevent invalid data from entering a dataset, so they would run on pre-commit phase. It would be very easy to add support for basic per-row SQL like checking for null or negative values, types that failed to parse/cast etc.

Checking for anomalies (e.g. sharp changes in daily case counts) is trickier. If we add such tests on a dataset that computes an aggregate - data that caused an anomaly is likely already in the system (in an upstream root dataset).

But we can start small - the goal as I see it is to reduce the amount of custom code & actions you need to support for your pipeline to function.

## Serving data
We are actively working on data APIs for `kamu`, e.g. last release has GraphQL support (`kamu system api-server --http-port 8080`) and lets you explore data and execute SQL queries on datasets. So in the long term you will be able to run `kamu` in an API server mode to provide a rich set of APIs to your users out of the box.

In short term it's probably not going to be the best use of your time to update your Flask app to **wrap** `kamu`'s API. 

We could start by simply exporting resulting datasets into Parquet, so you can keep using Pandas to load and query them with minimal or no changes.

## Manual data entry
Google Sheets can be directly ingested into `kamu` (via "Download as CSV" URL), so you could continue using them.

> Would also have to think about how historical data editing would work for datasets not provided in the form of time series

`kamu` can work with non-time-series data, but it will transform it into a ledger form using "Change Data Capture" (see [this example](https://docs.kamu.dev/cli/ingest/merge-strategies/#example----omit-in-toc----1)).
